### PR TITLE
Add the MeteringConfig resource to the metering client.

### PR DESCRIPTION
**Pull request by timflannagan1**
## To-Do:
- [x] Make Storage fields more general
- [x] Make TLS/auth fields more general
- [x] Add hadoop section
- [x] Add permissions section
- [x] Add openshift-reporting section
- [x] Verify the node selectors manifest works
- [x] Verify the default resource limits manifest works
- [x] Verify hdfs storage manifest works 
- [x] Verify latest version manifest works
- [x] Verify Metastore storage manifest works
- [x] Verify recommended resource limits works
- [x] Verify that the `omitempty` property is correct for all the boolean variables

## Things to Note:
- The Presto/Hive/Hadoop storage sections use a subset of the top-level storage AWS/Azure/GCS/S3Compatible/etc fields, but still reference those structures.
- Similarly to the storage reuse, some sections reference the `Jvm` structure but only use a subset of those values.
-  Need to be careful with regards to the `omitempty` property and structure fields that are a of boolean type, as `false` is considered an empty value in Go, but we don't necessarily want to discard that value (e.g. TLS configurations rely on `enabled: false`, `createSecret: false`).

### CODE REVIEW

File: pkg/apis/metering/v1/metering_config.go [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+	Route          *Route          `json:"route,omitempty"`
+	AuthProxy      *AuthProxy      `json:"authProxy,omitempty"`
+}
+type ReportingOperator struct {
+	Spec ReportingOperatorSpec `json:"spec"`
+}
+type Config struct {
+	AllNamespaces       bool                     `json:"allNamespaces,omitempty"`
+	EnableFinalizers    bool                     `json:"enableFinalizers,omitempty"`
+	LeaderLeaseDuration string                   `json:"leaderLeaseDuration,omitempty"`
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: This should be a `metav1.Duration` https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1

File: pkg/apis/metering/v1/metering_config.go [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+	Aws                 *AWS                     `json:"aws,omitempty"`
+	Hive                *ReportingOperatorHive   `json:"hive,omitempty"`
+	Presto              *ReportingOperatorPresto `json:"presto,omitempty"`
+	Prometheus          *Prometheus              `json:"prometheus,omitempty"`
+	TLS                 *APITLS                  `json:"tls,omitempty"`
+}
+
+type Affinity struct {
+}
+type Annotations struct {
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#ObjectMeta shows the types of annotations/labels which are `map[string]string`

File: pkg/apis/metering/v1/metering_config.go [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+	LogReports          bool                     `json:"logReports,omitempty"`
+	LogLevel            string                   `json:"logLevel,omitempty"`
+	Aws                 *AWS                     `json:"aws,omitempty"`
+	Hive                *ReportingOperatorHive   `json:"hive,omitempty"`
+	Presto              *ReportingOperatorPresto `json:"presto,omitempty"`
+	Prometheus          *Prometheus              `json:"prometheus,omitempty"`
+	TLS                 *APITLS                  `json:"tls,omitempty"`
+}
+
+type Affinity struct {
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: Is this needed?

File: pkg/apis/metering/v1/metering_config.go [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+	Values   []string `json:"values,omitempty"`
+}
+type LabelSelector struct {
+	MatchExpressions []MatchExpressions `json:"matchExpressions,omitempty"`
+}
+type RequiredDuringSchedulingIgnoredDuringExecution struct {
+	LabelSelector *LabelSelector `json:"labelSelector,omitempty"`
+	TopologyKey   string         `json:"topologyKey,omitempty"`
+}
+type PodAntiAffinity struct {
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: https://godoc.org/k8s.io/api/core/v1#PodAffinity

File: pkg/apis/metering/v1/metering_config.go [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+	RequiredDuringSchedulingIgnoredDuringExecution []RequiredDuringSchedulingIgnoredDuringExecution `json:"requiredDuringSchedulingIgnoredDuringExecution,omitempty"`
+}
+type PrestoAffinity struct {
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: No need for this type, you can use https://godoc.org/k8s.io/api/core/v1#Affinity

File: pkg/apis/metering/v1/metering_config.go [hunk 6]
```diff
@@ -0,0 +1,489 @@
...
+	MaxCachedBufferSize            int    `json:"maxCachedBufferSize,omitempty"`
+	MaxDirectMemorySize            int    `json:"maxDirectMemorySize,omitempty"`
+}
+type PrestoServerConfig struct {
+	Jvm                  *Jvm   `json:"jvm,omitempty"`
+	LogLevel             string `json:"logLevel,omitempty"`
+	TaskMaxWorkerThreads int    `json:"taskMaxWorkerThreads,omitempty"`
+	TaskMinDrivers       int    `json:"taskMinDrivers,omitempty"`
+}
+type PrestoLimits struct {
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: THese types aren't needed: https://godoc.org/k8s.io/api/core/v1#ResourceRequirements

File: pkg/apis/metering/v1/metering_config.go [hunk 7]
```diff
@@ -0,0 +1,489 @@
...
+	Requests Requests `json:"requests,omitempty"`
+}
+type Coordinator struct {
+	Affinity                      *PrestoAffinity     `json:"affinity,omitempty"`
+	NodeSelector                  *PrestoNodeSelector `json:"nodeSelector,omitempty"`
+	TerminationGracePeriodSeconds int                 `json:"terminationGracePeriodSeconds,omitempty"`
+	Config                        *PrestoServerConfig `json:"config,omitempty"`
+	Resources                     *PrestoResources    `json:"resources,omitempty"`
+}
+type PrestoSecurityContext struct {
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: https://godoc.org/k8s.io/api/core/v1#SecurityContext

File: pkg/apis/metering/v1/metering_config.go [hunk 8]
```diff
@@ -0,0 +1,489 @@
...
+	URL                  string                `json:"url,omitempty"`
+	CertificateAuthority *CertificateAuthority `json:"certificateAuthority,omitempty"`
+	MetricsImporter      *MetricsImporter      `json:"metricsImporter,omitempty"`
+}
+type PrometheusConfig struct {
+	ChunkSize                 string `json:"chunkSize,omitempty"`
+	PollInterval              string `json:"pollInterval,omitempty"`
+	StepSize                  string `json:"stepSize,omitempty"`
+	ImportFrom                string `json:"importFrom,omitempty"`
+	MaxImportBackfillDuration string `json:"maxImportBackfillDuration,omitempty"`
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#Duration

File: pkg/apis/metering/v1/metering_config.go [hunk 9]
```diff
@@ -0,0 +1,489 @@
...
+}
+
+type Prometheus struct {
+	URL                  string                `json:"url,omitempty"`
+	CertificateAuthority *CertificateAuthority `json:"certificateAuthority,omitempty"`
+	MetricsImporter      *MetricsImporter      `json:"metricsImporter,omitempty"`
+}
+type PrometheusConfig struct {
+	ChunkSize                 string `json:"chunkSize,omitempty"`
+	PollInterval              string `json:"pollInterval,omitempty"`
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#Duration

File: pkg/apis/metering/v1/metering_config.go [hunk 10]
```diff
@@ -0,0 +1,489 @@
...
+	SecretName   string `json:"secretName,omitempty"`
+}
+
+type Prometheus struct {
+	URL                  string                `json:"url,omitempty"`
+	CertificateAuthority *CertificateAuthority `json:"certificateAuthority,omitempty"`
+	MetricsImporter      *MetricsImporter      `json:"metricsImporter,omitempty"`
+}
+type PrometheusConfig struct {
+	ChunkSize                 string `json:"chunkSize,omitempty"`
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#Duration

File: pkg/apis/metering/v1/metering_config.go [hunk 11]
```diff
@@ -0,0 +1,489 @@
...
+	PollInterval              string `json:"pollInterval,omitempty"`
+	StepSize                  string `json:"stepSize,omitempty"`
+	ImportFrom                string `json:"importFrom,omitempty"`
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#Time

File: pkg/apis/metering/v1/metering_config.go [hunk 12]
```diff
@@ -0,0 +1,489 @@
...
+	MaxQueryLength                  string              `json:"maxQueryLength,omitempty"`
+	Aws                             *PrestoAWS          `json:"aws,omitempty"`
+	Azure                           *PrestoAzure        `json:"azure,omitempty"`
+	Gcs                             *PrestoGcs          `json:"gcs,omitempty"`
+	S3Compatible                    *PrestoS3Compatible `json:"s3Compatible"`
+	TLS                             *PrestoTLS          `json:"tls,omitempty"`
+	Auth                            *PrestoAuth         `json:"auth,omitempty"`
+	Connectors                      *Connectors         `json:"connectors,omitempty"`
+}
+type MatchExpressions struct {
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: Not needed

File: pkg/apis/metering/v1/metering_config.go [hunk 13]
```diff
@@ -0,0 +1,489 @@
...
+	Key      string   `json:"key,omitempty"`
+	Operator string   `json:"operator,omitempty"`
+	Values   []string `json:"values,omitempty"`
+}
+type LabelSelector struct {
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: Not needed

File: pkg/apis/metering/v1/metering_config.go [hunk 14]
```diff
@@ -0,0 +1,489 @@
...
+type PrestoHive struct {
+	HadoopConfigSecretName string `json:"hadoopConfigSecretName,omitempty"`
+	UseHadoopConfig        bool   `json:"useHadoopConfig,omitempty"`
+	MetastoreURI           string `json:"metastoreURI,omitempty"`
+	MetastoreTimeout       string `json:"metastoreTimeout,omitempty"`
+	TLS                    *TLS   `json:"tls,omitempty"`
+}
+type Connectors struct {
+	Hive PrestoHive `json:"hive,omitempty"`
+	// ExtraConnectorFiles []interface{} `json:"extraConnectorFiles,omitempty"`
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: the type for this is (uses an anonymous struct):
```
type ConnectorFileList []struct {
   Name string `json:"name"`
   Content string `json:"content"`
}
```

File: pkg/apis/metering/v1/metering_config.go [hunk 15]
```diff
@@ -0,0 +1,489 @@
...
+	StorageClass string `json:"storageClass,omitempty"`
+}
+
+/*
+End of storage section
+*/
+
+type PrestoLabels struct {
+}
+type PrestoImage struct {
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: This struct could be more generic and re-used probably

**Comment by timflannagan1** [comment]
Just pushed some changes before heading out, this should fail tests.

File: pkg/apis/metering/v1/metering_config.go (hunk 14 above)
Comment (timflannagan1) [reply pkg/apis/metering/v1/metering_config.go]: It seems like this needs to be a defined structure, as `make update-codegen` appears to fail anytime I define a (nested) anonymous structure.

Comment (chancez) [reply pkg/apis/metering/v1/metering_config.go]: That's fine.

File: pkg/apis/metering/v1/metering_config.go [hunk 16]
```diff
@@ -0,0 +1,633 @@
...
+type MeteringConfig struct {
+	meta.TypeMeta   `json:",inline"`
+	meta.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   MeteringConfigSpec   `json:"spec"`
+	Status MeteringConfigStatus `json:"status"`
+}
+
+type MeteringConfigStatus struct {
+	Storage             StorageSpec              `json:"storage"`
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: Probably should be a pointer, even if it's required.

File: pkg/apis/metering/v1/metering_config.go [hunk 17]
```diff
@@ -0,0 +1,633 @@
...
+	Hadoop              *Hadoop                  `json:"hadoop,omitempty"`
+
+	DisableOCPFeatures bool `json:"disableOCPFeatures"`
+	LogHelmTemplate    bool `json:"logHelmTemplate"`
+}
+
+type MeteringConfigSpec struct {
+	// Need to pass this as a pointer to a struct to avoid marshalling the zero-values despite omitempty being specified
+	// See: https://stackoverflow.com/questions/18088294/how-to-not-marshal-an-empty-struct-into-json-with-go
+	Storage             StorageSpec              `json:"storage"`
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: Pointer

File: pkg/apis/metering/v1/metering_config.go [hunk 18]
```diff
@@ -0,0 +1,633 @@
...
+	UnsupportedFeatures *UnsupportedFeaturesSpec `json:"unsupportedFeatures,omitempty"`
+	TLS                 *TLSSpec                 `json:"tls,omitempty"`
+	Monitoring          *MonitoringSpec          `json:"monitoring,omitempty"`
+	Permissions         *Permissions             `json:"permissions,omitempty"`
+	OpenshiftReporting  *OpenshiftReporting      `json:"openshift-reporting,omitempty"`
+	Ghostunnel          *Ghostunnel              `json:"__ghostunnel,omitempty"`
+	ReportingOperator   *ReportingOperator       `json:"reporting-operator,omitempty"`
+	Hadoop              *Hadoop                  `json:"hadoop,omitempty"`
+
+	DisableOCPFeatures bool `json:"disableOCPFeatures"`
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: omitEmpty

File: pkg/apis/metering/v1/metering_config.go [hunk 19]
```diff
@@ -0,0 +1,633 @@
...
+	LogHelmTemplate    bool `json:"logHelmTemplate"`
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: omitEmpty

File: pkg/apis/metering/v1/metering_config.go [hunk 20]
```diff
@@ -0,0 +1,633 @@
...
+	Image          *Image                       `json:"image,omitempty"`
+	UpdateStrategy *UpdateStrategy              `json:"updateStrategy,omitempty"`
+	Resources      *corev1.ResourceRequirements `json:"resources,omitempty"`
+	Route          *Route                       `json:"route,omitempty"`
+	AuthProxy      *AuthProxy                   `json:"authProxy,omitempty"`
+}
+type ReportingOperator struct {
+	Spec ReportingOperatorSpec `json:"spec,omitempty"`
+}
+type Config struct {
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: This should be `ReportingOperatorConfig`

File: pkg/apis/metering/v1/metering_config.go [hunk 21]
```diff
@@ -0,0 +1,633 @@
...
+	LogReports          bool                     `json:"logReports,omitempty"`
+	LogLevel            string                   `json:"logLevel,omitempty"`
+	Aws                 *AWS                     `json:"aws,omitempty"`
+	Hive                *ReportingOperatorHive   `json:"hive,omitempty"`
+	Presto              *ReportingOperatorPresto `json:"presto,omitempty"`
+	Prometheus          *Prometheus              `json:"prometheus,omitempty"`
+	TLS                 *APITLS                  `json:"tls,omitempty"`
+}
+
+type Labels struct {
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: Unneeded

File: pkg/apis/metering/v1/metering_config.go [hunk 22]
```diff
@@ -0,0 +1,633 @@
...
+
+type ReportingOperatorSpec struct {
+	Affinity       *corev1.Affinity             `json:"affinity,omitempty"`
+	Annotations    map[string]string            `json:"annotations,omitempty"`
+	Config         *Config                      `json:"config,omitempty"`
+	Labels         *Labels                      `json:"labels,omitempty"`
+	NodeSelector   *corev1.NodeSelector         `json:"nodeSelector,omitempty"`
+	Replicas       *int32                       `json:"replicas,omitempty"`
+	Image          *Image                       `json:"image,omitempty"`
+	UpdateStrategy *UpdateStrategy              `json:"updateStrategy,omitempty"`
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: This probably has a built-in type

File: pkg/apis/metering/v1/metering_config.go [hunk 23]
```diff
@@ -0,0 +1,633 @@
...
+	Aws                 *AWS                     `json:"aws,omitempty"`
+	Hive                *ReportingOperatorHive   `json:"hive,omitempty"`
+	Presto              *ReportingOperatorPresto `json:"presto,omitempty"`
+	Prometheus          *Prometheus              `json:"prometheus,omitempty"`
+	TLS                 *APITLS                  `json:"tls,omitempty"`
+}
+
+type Labels struct {
+}
+type AuthProxy struct {
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: Probably `ReportingOperatorAuthProxy`

File: pkg/apis/metering/v1/metering_config.go [hunk 24]
```diff
@@ -0,0 +1,633 @@
...
+	Enabled             bool                         `json:"enabled"`
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: omitEmpty

File: pkg/apis/metering/v1/metering_config.go [hunk 25]
```diff
@@ -0,0 +1,624 @@
...
+
+type ReportingOperator struct {
+	Spec ReportingOperatorSpec `json:"spec,omitempty"`
+}
+type ReportingOperatorSpec struct {
+	Affinity       *corev1.Affinity                  `json:"affinity,omitempty"`
+	Annotations    map[string]string                 `json:"annotations,omitempty"`
+	Config         *ReportingOperatorConfig          `json:"config,omitempty"`
+	Labels         *Labels                           `json:"labels,omitempty"`
+	NodeSelector   *corev1.NodeSelector              `json:"nodeSelector,omitempty"`
```
Comment (timflannagan1) [inline pkg/apis/metering/v1/metering_config.go]: @chancez we need to change this type to something other than `corev1.NodeSelector` or update the charts/manifests that handle the `nodeSelector` key as a list of key/value pairs.

Comment (timflannagan1) [reply pkg/apis/metering/v1/metering_config.go]: This is what is produced with the `corev1.NodeSelector` type:
```yaml
apiVersion: metering.openshift.io/v1
kind: MeteringConfig
...
  presto:
    spec:
      coordinator:
        nodeSelector:
          nodeSelectorTerms: null
      worker:
        nodeSelector:
          nodeSelectorTerms: null
...
```

File: pkg/apis/metering/v1/metering_config.go [hunk 26]
```diff
@@ -0,0 +1,624 @@
...
+	Replicas       *int32                            `json:"replicas,omitempty"`
+	Image          *Image                            `json:"image,omitempty"`
+	UpdateStrategy *appsv1.StatefulSetUpdateStrategy `json:"updateStrategy,omitempty"`
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: This should be a DeploymentUpdateStrategy

File: pkg/apis/metering/v1/metering_config.go [hunk 27]
```diff
@@ -67,22 +69,20 @@ type MeteringConfigSpec struct {
...
+type GenericTLSConfig struct {
 	Certificate   string `json:"certificate,omitempty"`
 	Key           string `json:"key,omitempty"`
 	CaCertificate string `json:"caCertificate,omitempty"`
 	CreateSecret  bool   `json:"createSecret"`
 	Enabled       bool   `json:"enabled"`
 	SecretName    string `json:"secretName,omitempty"`
 }
-type GeneralJVM struct {
+type GenericJVMConfig struct {
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: I think we could call this `JVMConfig`

File: pkg/apis/metering/v1/metering_config.go [hunk 28]
```diff
@@ -67,22 +69,20 @@ type MeteringConfigSpec struct {
...
-type Labels struct {
-}
-type Image struct {
+type GenericImageConfig struct {
 	PullPolicy string `json:"pullPolicy,omitempty"`
 	Repository string `json:"repository,omitempty"`
 	Tag        string `json:"tag,omitempty"`
 }
-type GeneralTLS struct {
+type GenericTLSConfig struct {
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: I think we could call this `TLSConfig`

File: pkg/apis/metering/v1/metering_config.go [hunk 29]
```diff
@@ -67,22 +69,20 @@ type MeteringConfigSpec struct {
 Start of structures re-used throughout the top-level keys
 */
 
-type Labels struct {
-}
-type Image struct {
+type GenericImageConfig struct {
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: I think we could call this `ImageConfig`

File: pkg/apis/metering/v1/metering_config.go [hunk 30]
```diff
@@ -0,0 +1,618 @@
...
+	Hadoop            *Hadoop            `json:"hadoop,omitempty"`
+	Presto            *Presto            `json:"presto,omitempty"`
+	ReportingOperator *ReportingOperator `json:"reporting-operator,omitempty"`
+}
+
+type MeteringConfigSpec struct {
+	DisableOCPFeatures bool `json:"disableOCPFeatures,omitempty"`
+	LogHelmTemplate    bool `json:"logHelmTemplate,omitempty"`
+
+	// Need to pass this as a pointer to a struct to avoid marshalling the zero-values despite omitempty being specified
```
Comment (chancez) [inline pkg/apis/metering/v1/metering_config.go]: Comment can be removed

**Pull request merged**